### ISSUE TITLE
Feedback - On Visual Scripting 

### ISSUE THREAD

**Issue by NinkuX** [godotengine/godot#12418]
_I am just copying and pasting feedback_ that was posted on a youtube video about Godots visual scripting. I think most artists, including myself feel the same about the current state of visual script. 

## Below is the feedback 
https://www.youtube.com/watch?v=WW8CJQ8jMBM&lc=z22ivh2ylynffdncaacdp432ibhs1vfmpnvp0o2qmsdw03c010c.1508858300040046

I feel like the programmers have missed the point of why artists and beginners want nodes in the first place. It's not just to have the code "be more visual". It's because typically in game engines, the nodes operate at a higher level allowing you to not have to know how the code works in as detailed a level or to go through so many steps to create interactions. For example, in the Blender game engine, I love the logic bricks because they allow me to get something working without having to know how to fully program in python. There is one block for just "move this object". I don't need to know all the functions etc to get the object, it's positions, create a variable, deal with deltas, add the variable to the position, update the object etc. I just use the "move this object" block... and it moves. Currently with Godot's visual scripting, it looks powerful, but it looks like I'd still need to know all of the programming, but just replace each line of code with a node. This basically gives all of the downsides of coding (need to have extensive knowledge and take multiple steps to perform even simple operations) with all of the downsides of nodes (slower to create each node than a line of code, quickly gets unreadable with complexity).

Devs seem to think that artists want visual scripting just because we inherently think in some way that makes us only able to understand pictures and not words. The real reason we want visual scripting is to make designing interactions simpler work less need to understand the underlying code with the same level of granularity.

I love Godot and I desperately want visual scripting, but as it is now, I don't see visual scripting speaking to artists or coders. Please get a group of the people this is targeted at and get their feedback at each step of the way. I guarantee that you'll see significant difference between how this is being designed and what they/we need.

Hopefully later we'll see some higher level nodes that combine a bunch of the current nodes together, eg a "move this object" node or a "spawn this object" node.
Now, to be sure there's a level of balance that needs to be stuck between how granular/powerful nodes are made vs how quick to create/easy they are. That needs to be based on the target audience. Currently though, this seems to be balanced to a target audience level of coders instead of artists.

Currently blender game engine leans a bit too far towards the simple/quick/limited end of the spectrum, while Godot is waaaaay too far towards the granular/flexible/complicated end of the spectrum.﻿

**Comment by reduz** [comment]
The aim of visual scripting in Godot is not to be like Construct or
GameMaker, but provide means for artists and level designers to have some
level of participation in the game logic.

It should be possible to provide something more akin to Blenderr using
custom nodes, but no one has made a proposal yet.

**Comment by gamergamesz** [comment]
well i think game developer can post there Custom node at godot asset library!
And i also fell like if godot VS workspace could be like blender Cycle node system in terms of working flexibility then it will be great for the user.

**Comment by Bugsbane** [comment]
As a technical artist, I get that we're not aiming at the construct/game maker/blender game engine crowd. I also understand that aiming a feature at everyone guarantees pleasing no one. At the same time, it seems to me that if artists need to understand coding to use visual scripting, that they're not going to use visual scripting anyway. To me, ideally, visual scripting would sit somewhere in-between game maker/construct/blender game maker and full on GDscript / python. Now I'm sure that replacing a line of code with a node is the easiest first implementation. If we're aiming visual scripting at *only* teams that have coders *and* artists who are technically advanced enough to be able to read visual scripts where a node is essentially a line of code, but aren't technically advanced enough to modify code directly though... that we're specializing down so much that visual scripting is unlikely to be useful to most people who would be interested in using it...

Whoever visual scripting is being aimed at though, let's aim to find a few people who fit that description and get feedback though. I saw this work very well for Krita when they adopted David Revoy as their test user. While I'd like to see Godot have more than one test user, it would be great to get feedback from users *in the target definition* to test assumptions and usefulness.

**Comment by benji011** [comment]
I'm a Python developer and I'm glad to see Godot introducing a feature to bring artists and coders together to work on game logic. I understand that is the purpose of 'visual scripting', in that you can 'see' the visual elements of the coding so there's no need to write a single line of code.

But, I also strongly agree in that the visual elements **isn't abstract/visual** enough for artists. 
I have a friend who worked a lot on Blender and i have heard him saying the exact same thing when he had a look at this.

Sure, an artist can learn. But the artist wouldn't be a real artist anymore if he has to worry about how a function should behave just to get it to move from one direction to another.

**Comment by NathanWarden** [comment]
I think we're far too close to release to change much with visual scripting now. I also think that the most useful feedback will come after Godot 3 is released. What I'm saying is that maybe Godot 3 will be the best place to get artist/designer feedback on it and adjust in Godot 4 where it will make sense to break some things, if indeed any breaking changes need to be made. Obviously if no breaking changes need to be made then these changes could be applied during the 3.x cycle.

**Comment by NullConstant** [comment]
_(I may be completely wrong on all this, as a disclaimer.)_

@NathanWarden IIRC a 3.0-tier breaking version update isn't really planned to happen for "Godot 4" (or planned at all considering that the furthest of plans don't reach beyond 3.2 at most currently), but that's beside the point.

I dabble equally in design, art, and programming, and while I can parse GDScript and VisualScript fine enough myself, I can see why it would be a bit daunting and/or not quite abstract enough for more specialized artists, but I think it's a problem that is best solved by custom, even more high-level nodes, built on top of the already existing system, ie. keeping the existing "VisualScript as visual GDScript" functionality while also adding extra stuff on top based on what artists and designers "in the real world" end up needing.

Perhaps it could be done through the ability to create custom nodes from GDScript? (or any other supported way like C# or GDNative bindings, though I think that's a bit of a corner case at that point)
I'm not quite sure about the specific things that'd be needed to make such an addition work, or if it's even the right way to approach this at all, though.

@reduz 

> The aim of visual scripting in Godot is not to be like Construct or GameMaker, but **provide means for artists and level designers to have some level of participation in the game logic**.

The question I have though, is why *would* artists/designers *want* to get themselves involved in the game logic side of things - barring situations in which a single person has to fill in multiple roles on a team (read: in a small indie studio), it just seems a bit weird and backwards.
Isn't it sort of the point of such systems, that the artists and designers can work on their creative vision without having to really getting down and dirty into how stuff really works deeper under the hood, beyond the high-level logic needed for them to actually implement what they want?
Because in its current state, VisualScript doesn't really accomplish that, or doesn't feel that way, at least, since it's overall so close to just being GDScript represented by node graphs rather than lines of code.

**Comment by insomniacUNDERSCORElemon** [comment]
I agree with what @Bugsbane says, especially on the scope only applying to teams resulting in most not liking the feature. I understand basic programming logic but don't have solid experience (save for 2 python-fu GIMP scripts), so is Godot's VS not intended for me?

VS could be so much more than it currently is:

- visual debugging for better view of where logic errors occur
- code collapsing that can be used to not only for better organization (without splitting everything into its own function) but custom pseudo-visual-nodes (control visible I/O, similar to script values that can be configured to be modified from the inspector)
- visual debugging+code collapsing could make design tutorials/examples much better
- more powerful search (and more included items, some of it contextual), so things like input and attribute editing (or anything else that you can't find directly in the node list) can be easily found... along with tooltips so you know what everything does without keeping the docs open permanently

Ultimately I think the biggest problem so far is that last point that makes it so it just isn't easy to start if you don't know the internals of GDscript, _even if you have experience with programming logic_. Ease of use probably should be the biggest priority, I can understand if "our aim with VS is to allow teamwork" is just for now... hopefully VS can be expanded later to the point where it is fit for (long-standing) individual use as well.

Further into the future, I think what would make VS good for artist-types on a team will be the same things making it better for _everyone_ interested in VS, the only thing (besides preference) that would really prevent usage in a full game would be performance (at least being lower than GDscript) or organization and both of those should be fixable.

**Comment by Bugsbane** [comment]
"it just isn't easy to start if you don't know the internals of GDscript, even if you have experience with programming logic."

Exactly.

Right now, it feels like users basically need to learn GDscript to set up a visual script, but that people who *do* know GDscript are unlikely to use the visual script system anyway (in its current form).

**Comment by reduz** [comment]
As I mentioned before, currently the visual scripting language _is_ limited to low level blocks and operations more akin to actual programming (in this way, it's similar to Unreal Blueprints). 

It is, however, possible to create custom and more high level blocks that may resemble more what you see in Game Maker or Blender logic bricks. As this logic is very "custom"  and requires the creation of higher level constructs that are not present anywhere in the engine, we simply don't ship Godot with that and never will.

Ideally, someone could create this as a downloadable module in the Asset Library, so It can be installed easily.

**Comment by insomniacUNDERSCORElemon** [comment]
@reduz , how about collapsing blocks/pseudo-VS-nodes (as I describe in my previous comment) that can be saved to be re-used? And then you have a select few of these (only ones that will be useful for most games) available to use, for instance detecting when a key is pressed to do something. Especially if these are modules or even JSON files (that can be pasted into a project).

I think this wouldn't really be "higher level" (even though it serves that purpose)... the actual activity of the logic isn't hidden (at least not any more than clicking an 'expand' button is hiding something), and it wouldn't be done inside the engine.

Obviously this would really only be for really basic stuff, and it'd use the structure of the 'best' way of doing it (basically the accepted way within the community). As this wouldn't cover much, there'd still be more to do to make VS more usable.

**Comment by vnen** [comment]
> The question I have though, is why would artists/designers want to get themselves involved in the game logic side of things - barring situations in which a single person has to fill in multiple roles on a team (read: in a small indie studio), it just seems a bit weird and backwards.
Isn't it sort of the point of such systems, that the artists and designers can work on their creative vision without having to really getting down and dirty into how stuff really works deeper under the hood, beyond the high-level logic needed for them to actually implement what they want?

> Ultimately I think the biggest problem so far is that last point that makes it so it just isn't easy to start if you don't know the internals of GDscript, *even if you have experience with programming logic*.

> Right now, it feels like users basically need to learn GDscript to set up a visual script, but that people who do know GDscript are unlikely to use the visual script system anyway (in its current form).

All of this misses how visual scripting is intended to be used in Godot (which is what driven its design). Visual scripts shouldn't be "made from scratch" by a designer. It should use custom nodes provided by the programmer. Better yet: the programmer already made the visual script with custom nodes and the designer can simply tweak it.

Visual script is not meant for the designer to understand the low level logic stuff, but the high level. Example: the player life bar blinks when it's below 20%. The designer can configure this value and the rate of blinking. The designer can also decide that if the player has a certain item, the life bar won't blink or will do so at a slower pace, then she can put a few conditional nodes in the script, based on information provided by custom nodes made by the programmer. This is something anyone can think of, it's not "programming logic". Of course, the designer needs to understand the basic visual scripting interface, how to use conditional blocks, etc. (but that's not what's being challenged).

----

As for general high-level nodes ready-to-use from the editor, I'm with @reduz: it can be provided in the Asset Library. With my proposal for better custom nodes (#9735, which I intend to work on), this should be easily accomplished.

**Comment by insomniacUNDERSCORElemon** [comment]
@vnen 
> Better yet: the programmer already made the visual script with custom nodes and the designer can simply tweak it.

That still goes along with what I said, but makes it even worse. That is such a niche use-case, if that's intended hardly anyone is going to use it.

I think VS as it is now wouldn't even be good enough for this. It sounds like it's really hand-holdy... like if a programmer made a VS for the life bar, I don't think a designer could sit down never using Godot's VS before and immediately _add_ the blinking action (guess I'd have to try that situation to know). If you're only talking about tweaking what the programmer has made, you might as well simply have them put everything in easy-to-understand variables and comments explaining what things do. It'd be better since they already need to write the code out and since they probably despise VS. And commenting is a good practice, anyways.

As I was trying to say with my previous comment, I don't think the issue is high vs low level... but that the issue is organization, tooltips, smart searching etc. that help users UNDERSTAND what they need to do. I think this is why Unreal's Blueprint system is actually usable.

By 'programming logic' I mean can the user think of at least pseudo-code to solve their issue. Like can you think of how to create a binary converter, how to prevent/fix logic issues, how to get the intended effect with conditions/math etc. Not everyone can think this way. Certainly someone with no programming experience (not even dabbling with simple scripting systems) is going to have more trouble with this than others. My point in that was this _isn't_ the issue, nor do I think it should-be-attempted-to/can be solved.

**Comment by reduz** [comment]
@insomniacUNDERSCORElemon As mentioned before, this is a first implementation. It has quite a handful users already and I'm sure it can be improved a lot so let's wait and see what people does with it and what feedback we get.

**Comment by Ranoller** [comment]
I come from the future... This can be fixed if VS-Script allows to extend gd files. Programmer write the script and the artist uses exported variables, custom functions, etc... In actual state, programmer will not want to use VS Script and artist neither.

**Comment by reduz** [comment]
Afaik you can create custom nodes easily in GDScript, but I don't think it was ever docummented....

**Comment by Ranoller** [comment]
The return of the jedi! ... We are waiting not for godot but yes for the godotcon report! 

Offtopic off: This documentation can push vs script to another state. A question: Why are vs files binary? Usually godot use plain text in files.

**Comment by mrcdk** [comment]
@reduz yes, it's possible to create custom nodes in GDScript. There are probably other ways to do it but the only way I think is useful is doing it in a plugin: [vs_extension.zip](https://github.com/godotengine/godot/files/1742257/vs_extension.zip)
